lib/evaluate/characteristics.cc

```diff
@@ -88,6 +84,16 @@ std::optional<TypeAndShape> TypeAndShape::Characterize(
   }
 }
 
+bool TypeAndShape::IsCompatibleWith(
+    parser::ContextualMessages &messages, const TypeAndShape &that) const {
+  if (!type_.IsTypeCompatibleWith(that.type_)) {
```

Comment:
Consider inverting this test and turning the body into an if/else structure, it might be more readable.

Reply:
OK